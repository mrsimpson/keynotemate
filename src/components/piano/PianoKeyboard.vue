<script lang="ts" setup>
import { computed } from 'vue'
import PianoKey from './PianoKey.vue'
import { pianoKeys } from '../../types/piano.ts'

const whiteKeys = computed(() => pianoKeys.filter((key) => !key.isBlack))
const blackKeys = computed(() => pianoKeys.filter((key) => key.isBlack))

// Calculate the total width of the white keys plus spacing
const whiteKeyWidth = 60
const whiteKeySpacing = 2
const totalWidth = computed(
  () => whiteKeys.value.length * (whiteKeyWidth + whiteKeySpacing) - whiteKeySpacing,
)

// Group black keys by offset and concatenate their names
const groupedBlackKeys = computed(() => {
  const groups: { [key: number]: { name: string; offset: number } } = {}
  blackKeys.value.forEach((key) => {
    if (key.offset !== undefined) {
      if (!groups[key.offset]) {
        groups[key.offset] = { name: key.name, offset: key.offset }
      } else {
        groups[key.offset].name += `/${key.name}`
      }
    }
  })
  return Object.values(groups)
})
</script>

<template>
  <div :style="{ width: `${totalWidth}px` }" class="piano-keyboard">
    <div v-for="octave in [4, 5]" :key="octave" class="keyboard-octave">
      <div class="keyboard-layout">
        <!-- White keys -->
        <div class="white-keys">
          <PianoKey
            v-for="key in whiteKeys"
            :key="key.name"
            :color="key.color"
            :isBlack="false"
            :name="key.name + octave"
            :octave="octave"
            :pitch="key.name"
          />
        </div>
        <!-- Black keys -->
        <div class="black-keys">
          <PianoKey
            v-for="key in groupedBlackKeys"
            :key="key.name"
            :isBlack="true"
            :name="key.name + octave"
            :octave="octave"
            :pitch="key.name.split('/')[0]"
            :style="{ left: `${key.offset}%` }"
            class="black-key-absolute"
          />
        </div>
      </div>
    </div>
  </div>
</template>

<style scoped>
.piano-keyboard {
  margin: 0 auto;
  padding: 20px;
}

.keyboard-layout {
  position: relative;
  display: flex;
  flex-direction: column;
}

.white-keys {
  display: flex;
  position: relative;
  z-index: 1;
  gap: 2px;
}

.black-keys {
  position: absolute;
  width: 100%;
  height: 0;
  z-index: 2;
}

.black-key-absolute {
  position: absolute;
  top: 0;
  transform: translateX(-30%);
}

/* Add borders to white keys */
.white-keys :deep(.white-key) {
  border: 1px solid #ccc;
  border-bottom-width: 4px;
}

/* Adjust black key positioning */
.black-keys :deep(.black-key) {
  border: 1px solid #000;
  border-bottom-width: 4px;
}
</style>
