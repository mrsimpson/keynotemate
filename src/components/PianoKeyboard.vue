<template>
  <div class="piano-keyboard">
    <div class="keyboard-layout">
      <!-- White keys -->
      <div class="white-keys">
        <PianoKey 
          v-for="key in whiteKeys" 
          :key="key.name"
          :name="key.name"
          :color="key.color"
          :isBlack="false"
        />
      </div>
      <!-- Black keys -->
      <div class="black-keys">
        <PianoKey 
          v-for="key in blackKeys" 
          :key="key.name"
          :name="key.name"
          :isBlack="true"
          :style="{ left: `${key.offset}%` }"
          class="black-key-absolute"
        />
      </div>
    </div>
  </div>
</template>

<script setup lang="ts">
import { computed, provide } from 'vue';
import PianoKey from './PianoKey.vue';
import { pianoKeys } from '../types/piano';
import { AudioService } from '../services/audioService';

const whiteKeys = computed(() => pianoKeys.filter(key => !key.isBlack));
const blackKeys = computed(() => pianoKeys.filter(key => key.isBlack));

// Create and provide the audio service
const audioService = new AudioService();
provide('audioService', audioService);
</script>

<style scoped>
.piano-keyboard {
  width: 100%;
  max-width: 700px;
  margin: 0 auto;
  padding: 20px;
}

.keyboard-layout {
  position: relative;
  display: flex;
  flex-direction: column;
}

.white-keys {
  display: flex;
  position: relative;
  z-index: 1;
  gap: 2px;
}

.black-keys {
  position: absolute;
  width: 100%;
  height: 0;
  z-index: 2;
}

.black-key-absolute {
  position: absolute;
  top: 0;
  transform: translateX(-50%);
}

/* Add borders to white keys */
.white-keys :deep(.white-key) {
  border: 1px solid #ccc;
  border-bottom-width: 4px;
}

/* Adjust black key positioning */
.black-keys :deep(.black-key) {
  border: 1px solid #000;
  border-bottom-width: 4px;
}
</style>
