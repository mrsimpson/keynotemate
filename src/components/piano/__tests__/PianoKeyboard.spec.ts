import { mount } from '@vue/test-utils'
import { describe, expect, it } from 'vitest'
import PianoKeyboard from '@/components/piano/PianoKeyboard.vue'

describe('PianoKeyboard Component', () => {
  it('renders default piano keyboard snapshot', () => {
    const wrapper = mount(PianoKeyboard)
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('renders keyboard with correct number of keys', () => {
    const wrapper = mount(PianoKeyboard)
    const whiteKeys = wrapper.findAll('.white-key')
    const blackKeys = wrapper.findAll('.black-key')

    expect(whiteKeys.length).toBe(14) // 7 white keys * 2 octaves
    expect(blackKeys.length).toBe(10) // 5 black keys * 2 octaves
    expect(wrapper.html()).toMatchSnapshot()
  })
})
