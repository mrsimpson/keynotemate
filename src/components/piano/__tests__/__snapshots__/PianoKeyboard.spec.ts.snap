// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PianoKeyboard Component > renders default piano keyboard snapshot 1`] = `
"<div data-v-32ce5279="" style="width: 432px;" class="piano-keyboard">
  <div data-v-32ce5279="" class="keyboard-octave">
    <div data-v-32ce5279="" class="keyboard-layout">
      <!-- White keys -->
      <div data-v-32ce5279="" class="white-keys">
        <div data-v-a7454f0c="" data-v-32ce5279="" class="white-key piano-key" style="background-color: rgb(255, 0, 0);">
          <div data-v-a7454f0c="" class="key-label">C4</div>
        </div>
        <div data-v-a7454f0c="" data-v-32ce5279="" class="white-key piano-key" style="background-color: rgb(255, 165, 0);">
          <div data-v-a7454f0c="" class="key-label">D4</div>
        </div>
        <div data-v-a7454f0c="" data-v-32ce5279="" class="white-key piano-key" style="background-color: rgb(255, 255, 0);">
          <div data-v-a7454f0c="" class="key-label">E4</div>
        </div>
        <div data-v-a7454f0c="" data-v-32ce5279="" class="white-key piano-key" style="background-color: rgb(0, 128, 0);">
          <div data-v-a7454f0c="" class="key-label">F4</div>
        </div>
        <div data-v-a7454f0c="" data-v-32ce5279="" class="white-key piano-key" style="background-color: rgb(0, 0, 255);">
          <div data-v-a7454f0c="" class="key-label">G4</div>
        </div>
        <div data-v-a7454f0c="" data-v-32ce5279="" class="white-key piano-key" style="background-color: rgb(75, 0, 130);">
          <div data-v-a7454f0c="" class="key-label">A4</div>
        </div>
        <div data-v-a7454f0c="" data-v-32ce5279="" class="white-key piano-key" style="background-color: rgb(148, 0, 211);">
          <div data-v-a7454f0c="" class="key-label">B4</div>
        </div>
      </div><!-- Black keys -->
      <div data-v-32ce5279="" class="black-keys">
        <div data-v-a7454f0c="" data-v-32ce5279="" class="black-key piano-key black-key-absolute" style="left: 12%;">
          <div data-v-a7454f0c="" class="key-label">C#/Db4</div>
        </div>
        <div data-v-a7454f0c="" data-v-32ce5279="" class="black-key piano-key black-key-absolute" style="left: 26%;">
          <div data-v-a7454f0c="" class="key-label">D#/Eb4</div>
        </div>
        <div data-v-a7454f0c="" data-v-32ce5279="" class="black-key piano-key black-key-absolute" style="left: 55%;">
          <div data-v-a7454f0c="" class="key-label">F#/Gb4</div>
        </div>
        <div data-v-a7454f0c="" data-v-32ce5279="" class="black-key piano-key black-key-absolute" style="left: 70%;">
          <div data-v-a7454f0c="" class="key-label">G#/Ab4</div>
        </div>
        <div data-v-a7454f0c="" data-v-32ce5279="" class="black-key piano-key black-key-absolute" style="left: 84%;">
          <div data-v-a7454f0c="" class="key-label">A#/Bb4</div>
        </div>
      </div>
    </div>
  </div>
  <div data-v-32ce5279="" class="keyboard-octave">
    <div data-v-32ce5279="" class="keyboard-layout">
      <!-- White keys -->
      <div data-v-32ce5279="" class="white-keys">
        <div data-v-a7454f0c="" data-v-32ce5279="" class="white-key piano-key" style="background-color: rgb(255, 0, 0);">
          <div data-v-a7454f0c="" class="key-label">C5</div>
        </div>
        <div data-v-a7454f0c="" data-v-32ce5279="" class="white-key piano-key" style="background-color: rgb(255, 165, 0);">
          <div data-v-a7454f0c="" class="key-label">D5</div>
        </div>
        <div data-v-a7454f0c="" data-v-32ce5279="" class="white-key piano-key" style="background-color: rgb(255, 255, 0);">
          <div data-v-a7454f0c="" class="key-label">E5</div>
        </div>
        <div data-v-a7454f0c="" data-v-32ce5279="" class="white-key piano-key" style="background-color: rgb(0, 128, 0);">
          <div data-v-a7454f0c="" class="key-label">F5</div>
        </div>
        <div data-v-a7454f0c="" data-v-32ce5279="" class="white-key piano-key" style="background-color: rgb(0, 0, 255);">
          <div data-v-a7454f0c="" class="key-label">G5</div>
        </div>
        <div data-v-a7454f0c="" data-v-32ce5279="" class="white-key piano-key" style="background-color: rgb(75, 0, 130);">
          <div data-v-a7454f0c="" class="key-label">A5</div>
        </div>
        <div data-v-a7454f0c="" data-v-32ce5279="" class="white-key piano-key" style="background-color: rgb(148, 0, 211);">
          <div data-v-a7454f0c="" class="key-label">B5</div>
        </div>
      </div><!-- Black keys -->
      <div data-v-32ce5279="" class="black-keys">
        <div data-v-a7454f0c="" data-v-32ce5279="" class="black-key piano-key black-key-absolute" style="left: 12%;">
          <div data-v-a7454f0c="" class="key-label">C#/Db5</div>
        </div>
        <div data-v-a7454f0c="" data-v-32ce5279="" class="black-key piano-key black-key-absolute" style="left: 26%;">
          <div data-v-a7454f0c="" class="key-label">D#/Eb5</div>
        </div>
        <div data-v-a7454f0c="" data-v-32ce5279="" class="black-key piano-key black-key-absolute" style="left: 55%;">
          <div data-v-a7454f0c="" class="key-label">F#/Gb5</div>
        </div>
        <div data-v-a7454f0c="" data-v-32ce5279="" class="black-key piano-key black-key-absolute" style="left: 70%;">
          <div data-v-a7454f0c="" class="key-label">G#/Ab5</div>
        </div>
        <div data-v-a7454f0c="" data-v-32ce5279="" class="black-key piano-key black-key-absolute" style="left: 84%;">
          <div data-v-a7454f0c="" class="key-label">A#/Bb5</div>
        </div>
      </div>
    </div>
  </div>
</div>"
`;

exports[`PianoKeyboard Component > renders keyboard with correct number of keys 1`] = `
"<div data-v-32ce5279="" style="width: 432px;" class="piano-keyboard">
  <div data-v-32ce5279="" class="keyboard-octave">
    <div data-v-32ce5279="" class="keyboard-layout">
      <!-- White keys -->
      <div data-v-32ce5279="" class="white-keys">
        <div data-v-a7454f0c="" data-v-32ce5279="" class="white-key piano-key" style="background-color: rgb(255, 0, 0);">
          <div data-v-a7454f0c="" class="key-label">C4</div>
        </div>
        <div data-v-a7454f0c="" data-v-32ce5279="" class="white-key piano-key" style="background-color: rgb(255, 165, 0);">
          <div data-v-a7454f0c="" class="key-label">D4</div>
        </div>
        <div data-v-a7454f0c="" data-v-32ce5279="" class="white-key piano-key" style="background-color: rgb(255, 255, 0);">
          <div data-v-a7454f0c="" class="key-label">E4</div>
        </div>
        <div data-v-a7454f0c="" data-v-32ce5279="" class="white-key piano-key" style="background-color: rgb(0, 128, 0);">
          <div data-v-a7454f0c="" class="key-label">F4</div>
        </div>
        <div data-v-a7454f0c="" data-v-32ce5279="" class="white-key piano-key" style="background-color: rgb(0, 0, 255);">
          <div data-v-a7454f0c="" class="key-label">G4</div>
        </div>
        <div data-v-a7454f0c="" data-v-32ce5279="" class="white-key piano-key" style="background-color: rgb(75, 0, 130);">
          <div data-v-a7454f0c="" class="key-label">A4</div>
        </div>
        <div data-v-a7454f0c="" data-v-32ce5279="" class="white-key piano-key" style="background-color: rgb(148, 0, 211);">
          <div data-v-a7454f0c="" class="key-label">B4</div>
        </div>
      </div><!-- Black keys -->
      <div data-v-32ce5279="" class="black-keys">
        <div data-v-a7454f0c="" data-v-32ce5279="" class="black-key piano-key black-key-absolute" style="left: 12%;">
          <div data-v-a7454f0c="" class="key-label">C#/Db4</div>
        </div>
        <div data-v-a7454f0c="" data-v-32ce5279="" class="black-key piano-key black-key-absolute" style="left: 26%;">
          <div data-v-a7454f0c="" class="key-label">D#/Eb4</div>
        </div>
        <div data-v-a7454f0c="" data-v-32ce5279="" class="black-key piano-key black-key-absolute" style="left: 55%;">
          <div data-v-a7454f0c="" class="key-label">F#/Gb4</div>
        </div>
        <div data-v-a7454f0c="" data-v-32ce5279="" class="black-key piano-key black-key-absolute" style="left: 70%;">
          <div data-v-a7454f0c="" class="key-label">G#/Ab4</div>
        </div>
        <div data-v-a7454f0c="" data-v-32ce5279="" class="black-key piano-key black-key-absolute" style="left: 84%;">
          <div data-v-a7454f0c="" class="key-label">A#/Bb4</div>
        </div>
      </div>
    </div>
  </div>
  <div data-v-32ce5279="" class="keyboard-octave">
    <div data-v-32ce5279="" class="keyboard-layout">
      <!-- White keys -->
      <div data-v-32ce5279="" class="white-keys">
        <div data-v-a7454f0c="" data-v-32ce5279="" class="white-key piano-key" style="background-color: rgb(255, 0, 0);">
          <div data-v-a7454f0c="" class="key-label">C5</div>
        </div>
        <div data-v-a7454f0c="" data-v-32ce5279="" class="white-key piano-key" style="background-color: rgb(255, 165, 0);">
          <div data-v-a7454f0c="" class="key-label">D5</div>
        </div>
        <div data-v-a7454f0c="" data-v-32ce5279="" class="white-key piano-key" style="background-color: rgb(255, 255, 0);">
          <div data-v-a7454f0c="" class="key-label">E5</div>
        </div>
        <div data-v-a7454f0c="" data-v-32ce5279="" class="white-key piano-key" style="background-color: rgb(0, 128, 0);">
          <div data-v-a7454f0c="" class="key-label">F5</div>
        </div>
        <div data-v-a7454f0c="" data-v-32ce5279="" class="white-key piano-key" style="background-color: rgb(0, 0, 255);">
          <div data-v-a7454f0c="" class="key-label">G5</div>
        </div>
        <div data-v-a7454f0c="" data-v-32ce5279="" class="white-key piano-key" style="background-color: rgb(75, 0, 130);">
          <div data-v-a7454f0c="" class="key-label">A5</div>
        </div>
        <div data-v-a7454f0c="" data-v-32ce5279="" class="white-key piano-key" style="background-color: rgb(148, 0, 211);">
          <div data-v-a7454f0c="" class="key-label">B5</div>
        </div>
      </div><!-- Black keys -->
      <div data-v-32ce5279="" class="black-keys">
        <div data-v-a7454f0c="" data-v-32ce5279="" class="black-key piano-key black-key-absolute" style="left: 12%;">
          <div data-v-a7454f0c="" class="key-label">C#/Db5</div>
        </div>
        <div data-v-a7454f0c="" data-v-32ce5279="" class="black-key piano-key black-key-absolute" style="left: 26%;">
          <div data-v-a7454f0c="" class="key-label">D#/Eb5</div>
        </div>
        <div data-v-a7454f0c="" data-v-32ce5279="" class="black-key piano-key black-key-absolute" style="left: 55%;">
          <div data-v-a7454f0c="" class="key-label">F#/Gb5</div>
        </div>
        <div data-v-a7454f0c="" data-v-32ce5279="" class="black-key piano-key black-key-absolute" style="left: 70%;">
          <div data-v-a7454f0c="" class="key-label">G#/Ab5</div>
        </div>
        <div data-v-a7454f0c="" data-v-32ce5279="" class="black-key piano-key black-key-absolute" style="left: 84%;">
          <div data-v-a7454f0c="" class="key-label">A#/Bb5</div>
        </div>
      </div>
    </div>
  </div>
</div>"
`;
